juju/model.py

```diff
@@ -0,0 +1,173 @@
+
+    def __repr__(self):
+        return f'<{type(self).__module__}.{type(self).__name__} {self.name}>'
+
+
+class Unit:
+    def __init__(self, name, cache):
+        self.name = name
+        self.app = cache.get(Application, name.split('/')[0])
+
+    def __repr__(self):
+        return f'<{type(self).__module__}.{type(self).__name__} {self.name}>'
+
+
+class LazyMapping(Mapping, ABC):
+    _lazy_data = None
+
+    @abstractmethod
+    def _load(self):
+        raise NotImplementedError()
+
+    @property
+    def _data(self):
+        data = self._lazy_data
+        if data is None:
+            data = self._lazy_data = self._load()
+        return data
+
+    def __contains__(self, key):
+        return key in self._data
+
+    def __len__(self):
+        return len(self._data)
+
+    def __iter__(self):
+        return iter(self._data)
+
+    def __getitem__(self, key):
+        return self._data[key]
+
+
+# TODO: Replace RelationMapping with non-lazy mapping with LazyList values instead.
+# Specifically, we want to avoid calling relation-ids for every relation and instead
+# only do it on demand for relations that are accessed. That will also allow us to
+# get rid of this extra layer of wrapping and just use a plain dict.
+class RelationMapping(LazyMapping):
+    """Map of relation names to lists of Relation instances."""
+    def __init__(self, relation_names, local_unit, backend, cache):
+        self._relation_names = relation_names
+        self._local_unit = local_unit
+        self._backend = backend
+        self._cache = cache
+
+    def _load(self):
+        data = {}
+        for relation_name in self._relation_names:
+            relations = data[relation_name] = []
+            for relation_id in self._backend.relation_ids(relation_name):
+                relations.append(Relation(relation_name, relation_id, self._local_unit, self._backend, self._cache))
+        return data
+
+
+class Relation:
+    def __init__(self, relation_name, relation_id, local_unit, backend, cache):
+        self.relation_name = relation_name
+        self.relation_id = relation_id
+        self._local_unit = local_unit
+        self._backend = backend
+        self._cache = cache
+        self._apps = None
+        self._units = None
+        self._data = None
+
+    @property
+    def apps(self):
+        if self._apps is None:
+            self._apps = list({unit.app for unit in self.units})
+        return self._apps
+
+    @property
+    def units(self):
+        if self._units is None:
+            self._units = []
+            for unit_name in self._backend.relation_list(self.relation_id):
+                self._units.append(self._cache.get(Unit, unit_name))
+        return self._units
+
+    @property
+    def data(self):
+        if self._data is None:
+            units = [self._local_unit] + self.units
+            self._data = {unit: RelationUnitData(self.relation_id, unit, self._backend) for unit in units}
+        return self._data
```

Comment:
Returning a full dictionary here feels a little bit fragile. Gut feeling is that something like what we had in the earlier versions of this PR, with a RelationData type that is more opaque and doesn't offer a traditional dictionary as its value, would be a more resilient design. But at this point it feels like we gain more by merging as-is and improving in a follow up.

Reply:
That will apply to the TODO comment around RelationMapping, as well, so we won't want to remove that wrapping when we do make it more on-demand.